File: library/src/main/java/wannabe/grid/OperationGrid.java
package wannabe.grid;

import java.util.Iterator;
import wannabe.Voxel;
import wannabe.grid.iterators.OperationIterator;
import wannabe.grid.operations.GroupOperation;
import wannabe.grid.operations.Operation;

/**
 * A Grid that caches the output of the specified operation(s) applied to the source Grid's data.
 */
public class OperationGrid implements Grid {
  private final Grid source;
  private final Operation op;

  public OperationGrid(Grid source, Operation operation) {
    this.source = source;
    op = operation;
  }

  public OperationGrid(Grid source, Operation... operations) {
    this.source = source;
    op = new GroupOperation(operations);
  }

  @Override public Iterator<Voxel> iterator() {
    return new OperationIterator(source.iterator(), op);
  }

  @Override public boolean isDirty() {
    return source.isDirty() && op.isDirty();
  }

  @Override public int size() {
    throw new IllegalStateException("Unable to determine size.");
  }

  @Override public AllNeighbors neighbors(Voxel voxel) {
    throw new IllegalStateException("Unable to determine neighbors.");
  }

  @Override public String toString() {
    return "op grid"; // TODO include op names or something
  }
}

File: library/src/main/java/wannabe/grid/OperationGrid.kt
package wannabe.grid

import wannabe.Voxel
import wannabe.grid.iterators.OperationIterator
import wannabe.grid.operations.GroupOperation
import wannabe.grid.operations.Operation

/**
 * A Grid that caches the output of the specified operation(s) applied to the source Grid's data.
 */
class OperationGrid(
  private val source: Grid,
  private val op: Operation
) : Grid {

  constructor(
    source: Grid,
    vararg operations: Operation
  ) : this(source, GroupOperation(*operations))

  override fun iterator(): Iterator<Voxel> {
    return OperationIterator(source.iterator(), op)
  }

  override val isDirty: Boolean
    get() = source.isDirty && op.isDirty

  override val size: Int
    get() = throw IllegalStateException("Unable to determine size.")

  override fun neighbors(voxel: Voxel): AllNeighbors =
    throw IllegalStateException("Unable to determine neighbors.")

  // TODO include op names or something
  override fun toString(): String = "op grid"
}
